metricflow/plan_conversion/node_processor.py

```diff
@@ -59,14 +61,111 @@ class MultiHopJoinCandidate:
     lineage: MultiHopJoinCandidateLineage
 
 
-@dataclass(frozen=True)
+class PredicatePushdownState(Enum):
+    """Enumeration of constraint states describing when predicate pushdown may operate in a dataflow plan.
+
+    This is necessary for holistic checks against the set of potentially enabled pushdown operations, because the
+    scenario where we only allow time range updates requires careful overriding of other pushdown properties.
+
+    Note: the time_range_only state is a backwards compatibility shim for use with conversion metrics while
+    we determine how best to support predicate pushdown for conversion metrics. It may have longer term utility,
+    but ideally we'd collapse this into a single enabled/disabled boolean property.
+    """
+
+    DISABLED = "disabled"
+    FULLY_ENABLED = "fully_enabled"
+    ENABLED_FOR_TIME_RANGE_ONLY = "time_range_only"
+
+
+@dataclasses.dataclass(frozen=True)
 class PredicatePushdownParameters:
-    """Container class for managing filter predicate pushdown.
+    """Container class for managing information about whether and how to do filter predicate pushdown.
 
-    Stores time constraint information for applying pre-join time filters.
+    The time_range_constraint property holds the time window for setting up a time range filter expression.
     """
 
-    time_range_constraint: Optional[TimeRangeConstraint]
+    _PREDICATE_METADATA_KEY = "is_predicate_property"
+
+    time_range_constraint: Optional[TimeRangeConstraint] = dataclasses.field(metadata={_PREDICATE_METADATA_KEY: True})
```

Comment:
This last was pretty hard to understand without diving into the dataclasses docs a bit & running the code to see what the value actually is. Might be helpful to leave a comment to explain that this is not a typical default arg (i.e., you still need to pass a value in for `time_range_constraint`)

Reply:
Yeah, to be honest I can probably get rid of this and just keep a list of predicate key names. That's what I had originally and I think it's a little less magical overall, but it's still a lot of magic string handling.